### DISCUSSION TITLE
Insights from Langchain Team

### DISCUSSION THREAD

**Discussion by lowspace** [lowspace/open_deep_research#24] [General]
Refers:

<a name="ft1">[1]</a>: https://blog.langchain.com/open-deep-research/
<a name="ft2">[2]</a>: https://rlancemartin.github.io/2025/07/30/bitter_lesson/
<a name="ft3">[3]</a>: https://www.youtube.com/watch?v=agGiWUpxkhg

**Comment by lowspace** [comment]
- #25 clarifier helps the system to fill the context gap between the human's intention and the real research target.
- #26 translator provides a standard data pre-process, ensuring the following agents to digest the research question without effort.
- #27 if a task requires highly consistent results, dividing it into parallel sub-tasks isn't promising—a hierarchical architecture would be better. Multi-agent systems are most beneficial for simple, parallel, and independent tasks.
- #28 multi-agent design => context isolation for the parallelized tasks
- #29 the research depth control mechanism is the same as the researcher agent, using the supervisor to control (soft) and a pre-defined parameter to stop (hard).
- context engineering => mitigate token bloat and steer behavior. 

The above is all based on https://blog.langchain.com/open-deep-research/

**Comment by lowspace** [comment]
在 _Learning the Bitter Lesson_ 里面分享了很多关于 open deep researcher 这个项目结构性以及未来前瞻的反思：

1. 关于结构（structure）中的假设

  > Often structure includes inductive biases about how we expect models to solve problems. Computer vision is a good example. For decades, researchers designed features (e.g., [SIFT](https://en.wikipedia.org/wiki/Scale-invariant_feature_transform) and [HOG](https://en.wikipedia.org/wiki/Histogram_of_oriented_gradients)) based upon domain knowledge. But these hand-crafted features restricted models to the patterns that we anticipated. As computation and data scaled, deep networks that learned features directly from pixels outperformed hand-crafted methods.

作者通过 CV feature engineering 从 domain knoweldge 过度到 pixels 的过程阐述 inductive bias 对工程或者模型的影响，以及新技术带来的改变。在 ML 时代，feature engineering 通常都是显式的，基于数学、统计、物理、专家知识等中高解释性模型得到的结果；在 DL 时代，ML 的做法被证明是低效的，专注于数据本身的表征（representation）是更高效的。NLP 也有类似的发展，从离散的特征空间，到连续的表征。

在此，作者想要说明一个核心观点：inductive biases 会影响 AI application 的构建，或者说 AI system 的结构构建。为了缓解 inductive biases 带来的影响，开发 llm system 的时候，需要进行一些「前瞻性」的思考，而非是单单「解决当下」问题。

> Jared Kaplan [(co-founder of Anthropic)](https://youtu.be/p8Jx4qvDoSo?si=giFZoWqhevhPn_qu) recently made the point that it can even be beneficial to “build things that don’t quite work” yet because the models will catch up (often quickly).

同时，需要尽可能透明化自己在构架 structure 时的 assumptions 或者是 motivations，这样在 llm capacity 提升的时候，能及时对此进行调整。

>I did not use tool calling, so I could not take advantage of the growing MCP ecosystem. The workflow always decomposed the request into report sections, a rigid research strategy that was not appropriate for all cases. The reports also were sometimes disjoint because I forced workers to write sections in parallel.


2. 如何更新结构？

> This is one of the main points of Hyung’s talk: we often [fail to remove](https://youtu.be/orDKvo8h71o?feature=shared&t=790) all the structure we add as we update our methods. In my case, I moved to an agent, but still was forcing each agent to write part of the report in parallel.

即使伴随着 llm 的发展或者业务的更新会对 agent structure 进行改变，某一些 module 的设计是能够继承下来的，比如 deep research 里面的「使用 sub-agent 完成平行、独立的任务」。

我认为这里的核心在于，某一些「机制」或者「机理」很有可能不会变化。比如 CoT 已经不会 explicitly 存在于 user prompt 里面了，但是它存在于 extended thinking 这个阶段中，内化为了 llm 的自然能力；比如 ReAct 这个行为框架在 agent 时代的 llm 已经被嵌入进训练过程中了<sup>[1](#ft1)</sup>，但是在工程化中，依然会使用 think_tool 这类工具进行 reasoning，同时在 workflow 设计中隐含 ReAct 框架。我认为这是这个机制是合理的：

1. extended thinking 是 overall thinking，针对是一整个 response，而 think_tool 是针对特定场景的反思，在设计上，某种程度实现了 context isolation，把 thinking 和 action 的 context 隔离开；

> 因为有 Agent 能力的模型，比如 Claude 4 系列（包括前面的 Claude 3.7 和 GPT-5），是不需要通过 ReAct 提示词来激发 Agent 能力，只要提供正确的工具和合适的工具描述，就会自动的去规划、调用工具和完成任务。<sup>[1](#ft1)</sup>

2. 这段话隐含了一个前提「llm 具备某种能力，那么就不太需要再次激发这种能力」。在 CoT 这个例子中，这个前提完全正确，对于 reasoning llm 的 prompt enigneering 已经完全不需要 CoT 来激发推理能力了。看起来拓展到工具调用上也很自然。有具体的反例了，JSON format 和 markdown format。即使 llm 经过了大量的 pre-training & post-training，想要输出稳定可靠的 JSON format，依然需要一些 few-shots；markdown format 的标准各异，想要输出一致性的 report，需要对特定的语法进行约定。所以这个前提不是 100% hold，而是需要具体情况具体分析的。在工具调用这个场景中，如果 llm 的作用是单纯的调用工具，不具备规划或者反思等能力，那么在 extended thinking 中能较为完美的达成「工具调用」这件事，但这里也有实现的缺陷：extended thinking 中不一定能 cover 特定的点，所以 prompt 中需要包含一些注意事项，保证 extended thinking 能执行。但现在 llm 的能力很强，一个 agent 的定位是单一的，但是功能是复合的，所以它会兼具研究、工具调用、反思者三个功能，这样单纯的 extended thinking 就缺乏针对性了。从这个角度来看，think_tool 衍生的 ReAct 设计是一个更巧妙的机制。

从以上论证可以衍生出结论：只要 agent 具备复合功能且 extened thinking 和 response 是分割开的，think_tool 这个 trick 在 agent 设计中应该能一直存在。

微博的这句话是对的吗？不完全对。

- 「扔掉 ReAcT 模版」，没问题；
- 需要「提供正确的工具 + 合适的工具表达」；
- 「正确的工具 + 合适的工具表达」!= 「自动规划和调用的能力是令人满意的」。

更准确的表达应该是，现在可以扔掉 ReAct prompt 模版，转而向模型提供特定的工具来实现 ReAct 框架，使得模型能对其它工具进行自动规划、调用、和评估。

这个例子给了更新结构的启示：

- 需要根据 llm update 评估结构 => 判断是否需要修改；
- 在代码和维护层面，结构的实现需要做成「便于拆卸的模块」，随时更改 => 类似于使用 torch 搭 nn layer 的感觉，对齐 input 和 output 之后，中间的 module 能随意替换。

<a name="ft1">[1]</a>: https://weibo.com/1727858283/Q4mk9orzU
<a name="ft2">[2]</a>: https://rlancemartin.github.io/2025/07/30/bitter_lesson/
